pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java

```diff
@@ -0,0 +1,168 @@
+
+  private final RangeBitmap.Appender _appender;
+  private final File _rangeIndexFile;
+  private final FieldSpec.DataType _valueType;
+  private final long _minValue;
+  private final double _scale;
+
+  public BitSlicedRangeIndexCreator(File indexDir, ColumnMetadata metadata) {
+    _appender = RangeBitmap.appender(maxValue(metadata));
+    _rangeIndexFile = new File(indexDir, metadata.getColumnName() + BITMAP_RANGE_INDEX_FILE_EXTENSION);
+    _valueType = metadata.getDataType();
+    _minValue = minValue(metadata);
+    _scale = scale(metadata);
+  }
+
+  @Override
+  public void add(int dictId) {
+    _appender.add(dictId - _minValue);
+  }
+
+  @Override
+  public void add(int[] dictIds, int length) {
+  }
+
+  @Override
+  public void add(long value) {
+    _appender.add(value - _minValue);
+  }
+
+  @Override
+  public void add(long[] values, int length) {
+  }
+
+  @Override
+  public void add(float value) {
+    if (!Float.isFinite(value)) {
+      // TODO this probably needs better handling, like using a 24th slice
+      //  for infinite values
+      _appender.add(0);
+    } else {
+      double d = value;
+      d -= Double.longBitsToDouble(_minValue);
+      d *= _scale;
+      _appender.add(Float.floatToIntBits((float) d));
+    }
+  }
+
+  @Override
+  public void add(float[] values, int length) {
+  }
+
+  @Override
+  public void add(double value) {
+    if (!Double.isFinite(value)) {
+      // TODO this probably needs better handling, like using a 57th slice
+      //  for infinite values
+      _appender.add(0L);
+    } else {
+      value -= Double.longBitsToDouble(_minValue);
+      value *= _scale;
+      _appender.add(Double.doubleToLongBits(value));
+    }
+  }
+
+  @Override
+  public void add(double[] values, int length) {
+  }
+
+  @Override
+  public void seal() throws IOException {
+    int headerSize = Integer.BYTES + Long.BYTES + (_valueType == FLOAT || _valueType == DOUBLE ? Double.BYTES : 0);
+    int serializedSize = _appender.serializedSizeInBytes();
+    try (MmapFileWriter writer = new MmapFileWriter(_rangeIndexFile, headerSize + serializedSize)) {
+      writer.write(buf -> {
+        buf.putInt(RANGE_INDEX_VERSION);
+        buf.putLong(_minValue);
+        if (_valueType == FLOAT || _valueType == DOUBLE) {
+          buf.putDouble(_scale);
+        }
+        _appender.serialize(buf);
+      });
+    }
+    _appender.clear();
+  }
+
+  @Override
+  public void close() throws IOException {
+
+  }
+
+  private static long maxValue(ColumnMetadata metadata) {
+    FieldSpec.DataType dataType = metadata.getDataType();
+    if (dataType == LONG || dataType == INT) {
+      Comparable<?> minValue = metadata.getMinValue();
+      Comparable<?> maxValue = metadata.getMaxValue();
+      if (minValue instanceof Number && maxValue instanceof Number) {
+        return ((Number) maxValue).longValue() - ((Number) minValue).longValue();
+      }
+    }
+    if (dataType == DOUBLE) {
```

Comment:
Putting some comments on how it works would be good

Reply:
This is a WIP and will change, but I will be sure to document the mapping clearly when it is ready.

Reply:
This is encapsulated in the class `FPOrdering` now.